servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,421 @@
+    if (lineage == null) {
+      throw new LineageDoesNotExistException(
+          "the lineage " + lineageId + " to delete does not exist");
+    }
+
+    // there should not be child lineage if not cascade
+    if (!cascade && !mLineageStore.getChildren(lineage).isEmpty()) {
+      throw new LineageDeletionException(
+          "the lineage " + lineageId + " to delete has children lineages");
+    }
+
+    LOG.info("Delete lineage " + lineageId);
+    mLineageStore.deleteLineage(lineageId);
+    return true;
+  }
+
+  /**
+   * Reinitializes the file when the file is lost or not completed.
+   *
+   * @param path the path to the file
+   * @param blockSizeBytes the block size
+   * @param ttl the TTL
+   * @return the id of the reinitialized file when the file is lost or not completed, -1 otherwise
+   * @throws InvalidPathException the file path is invalid
+   */
+  public long recreateFile(String path, long blockSizeBytes, long ttl)
+      throws InvalidPathException, LineageDoesNotExistException {
+    long fileId = mFileSystemMaster.getFileId(new TachyonURI(path));
+    LineageFileState state = mLineageStore.getLineageFileState(fileId);
+    if (state == LineageFileState.CREATED || state == LineageFileState.LOST) {
+      LOG.info("Recreate the file " + path + " with block size of " + blockSizeBytes + " bytes");
+      return mFileSystemMaster.reinitializeFile(new TachyonURI(path), blockSizeBytes, ttl);
+    }
+    return -1;
+  }
+
+  public void asyncCompleteFile(long fileId)
+      throws FileDoesNotExistException, BlockInfoException {
+    LOG.info("Async complete file " + fileId);
+    mLineageStore.completeFile(fileId);
+    // complete file in Tachyon.
+    mFileSystemMaster.completeFile(fileId);
+    writeJournalEntry(new AsyncCompleteFileEntry(fileId));
+    flushJournal();
+  }
+
+  private void asyncCompleteFileFromEntry(AsyncCompleteFileEntry entry) {
+    mLineageStore.completeFile(entry.getFileId());
+  }
+
+  /**
+   * Instructs a worker to persist the files for checkpoint.
+   *
+   * TODO(yupeng) run the heartbeat in a thread?
+   *
+   * @param workerId the id of the worker that heartbeats
+   * @return the command for checkpointing the blocks of a file
+   * @throws FileDoesNotExistException if the file does not exist
+   */
+  public LineageCommand lineageWorkerHeartbeat(long workerId, List<Long> persistedFiles)
+      throws FileDoesNotExistException {
+    // notify checkpoint manager the persisted files
+    persistFiles(workerId, persistedFiles);
+
+    // get the files for the given worker to checkpoint
+    List<CheckpointFile> filesToCheckpoint = null;
+    filesToCheckpoint = pollToCheckpoint(workerId);
+    if (!filesToCheckpoint.isEmpty()) {
+      LOG.info("Sent files " + filesToCheckpoint + " to worker " + workerId + " to persist");
+    }
+    return new LineageCommand(CommandType.Persist, filesToCheckpoint);
+  }
+
+  public List<LineageInfo> listLineages() {
```

Comment:
I think it should be synchronized, because what if structure changes are made to the Lineage during the process of getting LineageInfo?